experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+          continue;
+        coveredPredecessors.insert(pred);
+
+        // Get the branch terminator
+        auto branchOp = dyn_cast<BranchOpInterface>(pred->getTerminator());
+        assert(branchOp && "Expected terminator operation in a predecessor "
+                           "block feeding a block argument!");
+
+        // Check if the input value `c` of type Value is already present among
+        // the operands of the phi function
+        auto isAlreadyPresent = [&](Value c) -> bool {
+          return std::any_of(operands.begin(), operands.end(),
+                             [c](GateInput *in) {
+                               return in->isTypeValue() && in->getValue() == c;
+                             });
+        };
+
+        // For each alternative in the branch terminator
+        for (auto [successorId, successorBlock] :
+             llvm::enumerate(branchOp->getSuccessors())) {
+
+          // Skip the successor if it is not the block under analysis
+          if (successorBlock != &block)
+            continue;
+
+          // Get the values used for that branch
+          auto successorOperands = branchOp.getSuccessorOperands(successorId);
+          // Get the value used as input of the gate
+          auto producer = successorOperands[argNumber];
+          GateInput *gateInput = nullptr;
+
+          /// If it is a block argument whose parent block has some predecessor,
+          /// then the value is the output of a phi, and we add it to the list
+          /// of missing phis. Otherwise, the input is a value, and it can be
+          /// safely added directly.
+          if (auto ba = dyn_cast<BlockArgument>(producer);
+              ba && !producer.getParentBlock()->hasNoPredecessors()) {
+            gateInput = new GateInput((Gate *)nullptr);
+            phisToConnect.push_back(MissingPhi(gateInput, ba));
+          } else {
+            if (!isAlreadyPresent(dyn_cast<Value>(producer)))
+              gateInput = new GateInput(producer);
+          }
+
+          // Insert the value among the inputs of the phi
+          if (gateInput)
+            operands.push_back(gateInput);
+
+          break;
+        }
+      }
+
+      // If the list of operands is not empty (i.e. the phi has at least
+      // one input), add it to the phis associated to that block
+      if (!operands.empty()) {
+        auto *newPhi =
+            new Gate(arg, operands, GateType::PhiGate, ++uniqueGateIndex);
+        gateList[&block].push_back(newPhi);
+      }
+    }
+  }
+
+  // Find the missing phi and correct the pointers
+  for (MissingPhi &missing : phisToConnect) {
+    auto list = gateList[missing.ba.getParentBlock()];
+    auto foundGate = std::find_if(list.begin(), list.end(),
+                                  [&](auto &t) {
+                                    return t->getArgumentNumber() ==
+                                           missing.ba.getArgNumber();
+                                  }
```

Comment:
The fact that you have to do a linear search suggest this should be some sort of set, possibly with a deterministic iteration order if that is important.

Reply:
Sorry my brain goofed and I miss-read this, it's actually fine. My bad!